src/python_testing/mdns_discovery/mdns_discovery.py

```diff
@@ -0,0 +1,275 @@
+#
+#    Copyright (c) 2023 Project CHIP Authors
+#    All rights reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+
+import asyncio
+from typing import Dict, List
+
+from mdns_discovery.exceptions import DiscoveryNotPerformedError
+from zeroconf import ServiceStateChange, Zeroconf
+from zeroconf.asyncio import AsyncServiceBrowser, AsyncServiceInfo
+
+from .mdns_type_enum import MdnsType
+
+
+class MdnsDiscovery:
```

Comment:
How was this tested for compliance with DNS-SD/mDNS RFCs? Was it observed that *all* AAAA records were returned for multi-address devices rather than just one? Is it compliant with testing mDNS only over IPv6 to avoid reliance on IPv4 behaviors that re not specified? This will be used in cert tests so it has to be as compliant as possible

Reply:
This class is based on the **zeroconf** module to perform the services discovery.

The relevant RFCs used are mentioned throughout the [zeroconf ChangeLog](https://github.com/python-zeroconf/python-zeroconf/blob/master/CHANGELOG.md) (RFC 6762, RFC 1035, RFC 4007)

Zeroconf can be used in IPv6 only mode like this:
`Zeroconf(ip_version=IPVersion.V6Only)`

I will make that change.

Reply:
I looked at the docs, and it doesn't appear that the interface is listed in IPv6 addresses for Link-Local addresses to be correctly supported.

Did you test this modules with IPv6 devices that have multiple addresses, including link-local addresses?

Reply:
We don't have tests that require link local addresses on AAAA because the spec doesn't mandate that. This gives us a way to check for the validity of txt records and instance/service names etc, all of which are now currently manual. If we want to check for mdns stack interoperability, we're better off doing that in the interop lab, where we get both sides of the stack. 

If the spec wants to mandate that certain addresses appear in the the quad-As, then it needs to be adjusted and we can add tests for such things. Right now, it's just "all available IPv6 addresses upon which they are willing to accept operational messages". We know from other tests that at least 1 needs to work. It's not guaranteed to be the link local. The ordering in the sdk is heuristic.

Reply:
The ordering in the SDK prefers link-local always. The question I asked is based on a lot of experience with actual Matter devices in our ecosystem in the field. If we are going to start automating mDNS testing, which we should, then we should make sure we are as compliant as possible on the cert/test side. I am not blocking the start of this effort, I am asking legitimate technical questions about a deliverable being provided for ostensibly improving certification tests quality.

Reply:
I've opened an issue to track the testing improvements we want to see for mdns here: https://github.com/CHIP-Specifications/chip-test-plans/issues/3926. Please add as required to the issue.

For this test, I want zeroconf (well, not specifically zeroconf, but any third party) mdns tool used so we're not verifying the SDK against the SDK.

Going to close this for now so we can get this in to start the automation process, please add any additional comments on the issue.